tensornetwork/backends/base_backend.py

```diff
@@ -135,6 +135,19 @@ def shape(self, tensor: Tensor) -> Tensor:
     raise NotImplementedError("Backend '{}' has not implemented shape.".format(
         self.name))
 
+  def tuple_shape(self, tensor: Tensor) -> Tensor:
```

Comment:
We do need both.

We need the `shape` method so that we can do the correct shape calculations when given tensorflow tensors with undefined dimensions. So in this case, `backend.shape` actually returns a `Tensor` type. Whereas `shape_tuple` will return a a tuple of optional ints. This shape will be used for things where the shape will change the python runtime branching (which can't be done if the shape is a tensor type). 

It's a weird and confusing difference, but it has to be done simply because tensorflow tensors can have undefined dimension sizes.